### ISSUE TITLE
Use too much memory, how to limit it?

### ISSUE THREAD

**Issue by zhoulukuan** [NVIDIA/DALI#1198]
At first I used the 16G memory running program, the memory usage continued to rise, and finally the program crashed. After using different machines with 50G and 200G memory, using dali also takes up a lot of memory(about 100% and 87% on this two machine). How to avoid this phenomenon? How to limit it？Is it because I use too many threads?

**Comment by awolant** [comment]
Hi, thanks for the question.
Are you seeing this rise in memory usage during the initial phase of the training? If so, that is expected. DALI uses memory buffers for outputs of every operator. If you input images vary in size, those buffers will be expanded to accommodate the biggest images. After a while memory usage should stabilize. 
Memory usage is also very dependent on the exact pipeline you are using. If you have a lot of operators it can get quite high. Could you post some code snippets for us to take a look at?

**Comment by zhoulukuan** [comment]
@awolant I use dali to read dataset from rec and eval images with pytorch. Reading code:
```
from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops
import nvidia.dali.types as types
from nvidia.dali.plugin.pytorch import DALIGenericIterator, DALIClassificationIterator

class CommonPipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, mode, size, dataset):
        super(CommonPipeline, self).__init__(batch_size, num_threads, device_id)
        self.decode = ops.nvJPEGDecoder(device="mixed", output_type=types.RGB)
        self.mode = mode
        self.resize = ops.Resize(device="gpu",
                                 image_type=types.RGB,
                                 interp_type=types.INTERP_LINEAR)
        self.resize_rng = ops.Uniform(range=size)
        self.flip = ops.Flip(device="gpu")
        self.coin = ops.CoinFlip(probability=0.5) if dataset == 'train' else ops.CoinFlip(probability=0)
        self.cmn = ops.CropMirrorNormalize(device="gpu",
                                            output_dtype=types.FLOAT,
                                            crop=(224, 224),
                                            image_type=types.RGB,
                                            mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
                                            std=[0.229 * 255, 0.224 * 255, 0.225 * 255])
        self.uniform = ops.Uniform(range=(0.0, 1.0))

    def base_define_graph(self, inputs, labels):
        origins = self.decode(inputs)
        if self.mode == 'shorter':
            images = self.resize(origins, resize_shorter=self.resize_rng())
        else:
            images = self.resize(origins, resize_x=self.resize_rng(), resize_y=self.resize_rng())
        images = self.flip(images, horizontal=self.coin())
        output = self.cmn(images, crop_pos_x = self.uniform(),
                          crop_pos_y = self.uniform())
        return (output, labels)


class MXNetReaderPipeline(CommonPipeline):
    def __init__(self, folder, batch_size, num_threads, shuffle, mode, size, dataset, device_id, num_gpus):
        super(MXNetReaderPipeline, self).__init__(batch_size, num_threads, device_id, mode, size, dataset)
        self.input = ops.MXNetReader(path=[folder+"temp.rec"], index_path=[folder+"temp.idx"],
                                     random_shuffle=shuffle, shard_id=device_id, num_shards=num_gpus)

    def define_graph(self):
        images, labels = self.input(name="Reader")
        return self.base_define_graph(images, labels)


def get_dataset_from_rec(db_folder, batch_size, gpus, shuffle, mode, size, num_threads, dataset):
    N = len(gpus)
    pipes = [MXNetReaderPipeline(folder=db_folder, batch_size=batch_size, num_threads=num_threads, shuffle=shuffle,
                                 mode=mode, size=size, dataset=dataset,
                                 device_id=device_id, num_gpus=N) for device_id in gpus]
    pipes[0].build()
    # return DALIGenericIterator(pipes, ['data', 'label'], pipes[0].epoch_size("Reader"))
    return DALIGenericIterator(pipes, ['data', 'label'], pipes[0].epoch_size("Reader")), pipes[0].epoch_size("Reader")
```
In the testing program, I read different models and calculate their performance on this rec dataset. Each iteration process model will load new parameters, and at the end I will reset loader

**Comment by zhoulukuan** [comment]
About how to get dataloader:
```
loader, num = get_dataset_from_rec(db_folder=args.rec, gpus=[0], batch_size=args.batch_size,
                        shuffle=False, mode='both', size=224, num_threads=8, dataset="test")
```
When the memory is 16g, the memory usage will start to increase rapidly, then the growth rate  starts to slow down. The overall usage began to increase slowly, the whole process lasts for 6~7 hours and then the program crashes.

**Comment by awolant** [comment]
Ok, for this pipeline memory consumption you are reporting feels a bit off. What dataset exactly are you using? Is it publicly available?

**Comment by zhoulukuan** [comment]
Sorry, I am using data that is not publicly available. However, many of these data are collected online(such as Google search), so the format is very diverse(jpg/jpeg/png/gif).

**Comment by zhoulukuan** [comment]
I do some experiment and found about 50g memory could supprt the program runs successfully util it ends. It is normal?

**Comment by mzient** [comment]
CPU operators (such as readers) keep a separate buffer for each sample. These buffers are bound to a particular sample index in the batch. If a particularly large sample appears at multiple indices in a batch due to shuffling between epochs, then there will be multiple very large buffers.
GPU operators follow a different strategy, but from your initial post I understood that it's the CPU memory that's problematic.
BTW - which version of DALI are you using? We've changed host memory handling recently, so maybe there's some new problem which we're not aware of.

**Comment by zhoulukuan** [comment]
@mzient dali version: 0.12.0 Python: 2.7.16
Do you mean the buffers should be fixed, but between epochs, more multiple large buffers are create? How to avoid it?

**Comment by awolant** [comment]
There is no way to avoid it. It's just how DALI works. If your data set has big images, they will land in different buffers during the training. That is because they will be at the different indexes within the batch. To accommodate them, buffers will be increased. That is why memory usage increases during the training, especially in the beginning.
Without more information about your data set it is hard to tell, if what you are experiencing is a memory leak of some sorts or is it just normal behavior given your data.
Could you tell us what sizes are the images? What is the variance in sizes in your data set? Are there any images way bigger then the rest?

**Comment by zhoulukuan** [comment]
The dataset file is about 132GB in size and contains approximately 180,000 images. Most of these images are collected from the Internet and are therefore very different in size and aspect ratio. 
So it’s likely to have some images with large aspect ratios or large sizes than others.

**Comment by awolant** [comment]
Right. As we wrote before, large images can significantly increase memory usage in DALI. Since you collected them from the internet I suspect some of them might be quite big in size.
Maybe you could try to filter out the biggest images and see how your code behaves?

**Comment by zhoulukuan** [comment]
OK, I will have a try.

**Comment by mzient** [comment]
We're pleased to say that we've changed the allocation strategy for (non-pinned) CPU buffers. It reduces the memory consumption in RN50 training in PyTorch by almost 50%. Please check the latest master (or next successul nightly) and see if your issue is resolved.
The memory is now freed when a requested tensor is smaller than a given percentage of actual allocation. You can tweak it by setting the environment variable `DALI_HOST_BUFFER_SHRINK_THRESHOLD=0.xx`. The default value is 0.9. You can also set it in python using ``nvidia.dali.backend.SetHostBufferShrinkThreshold(threshold)`

**Comment by JanuszL** [comment]
0.19 is out and should address this. Please reopen if it doesn't work.
